java/core/src/main/java/b3/TraceContext.java

```diff
@@ -0,0 +1,355 @@
+ * representation.
+ *
+ * <p>This binary representation is fixed-length, depending on the size of the trace identifier.
+ * This representation doesn't have a way to represent an absent parent (root span). In this
+ * serialized form, a root span is when all three ids are the same. Alternatively, you can use
+ * {@link #nullableParentId}.
+ *
+ * <p>The implementation was originally {@code com.github.kristofa.brave.TraceContext}, which was a
+ * port of {@code com.twitter.finagle.tracing.TraceId}.
+ */
+public final class TraceContext {
+
+  /**
+   * Unique 8-byte or 16-byte identifier for a trace, set on all spans within it.
+   */
+  public static final class TraceId {
+    public static TraceId fromLong(long lo) {
+      return new TraceId(0L, lo);
+    }
+
+    /** 0 may imply 8-byte identifiers are in use */
+    public final long hi;
+    public final long lo;
+
+    TraceId(long hi, long lo) {
+      this.hi = hi;
+      this.lo = lo;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) return true;
+      if (o instanceof TraceId) {
+        TraceId that = (TraceId) o;
+        return (this.hi == that.hi)
+            && (this.lo == that.lo);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      int h = 1;
+      h *= 1000003;
+      h ^= (hi >>> 32) ^ hi;
+      h *= 1000003;
+      h ^= (lo >>> 32) ^ lo;
+      return h;
+    }
+  }
+
+  public static final int FLAG_DEBUG = 1 << 0;
+  /** When set, we can interpret {@link #FLAG_SAMPLED} as a set value. */
+  public static final int FLAG_SAMPLING_SET = 1 << 1;
+  public static final int FLAG_SAMPLED = 1 << 2;
+  /**
+   * When set, we can ignore the value of the {@link #parentId}
+   *
+   * <p>While many zipkin systems re-use a trace id as the root span id, we know that some don't.
+   * With this flag, we can tell for sure if the span is root as opposed to the convention trace id
+   * == span id == parent id.
+   */
+  public static final int FLAG_IS_ROOT = 1 << 3;
+
+  TraceContext(TraceId traceId, long parentId, long spanId, long flags) {
+    this.traceId = traceId;
+    this.parentId = (parentId == spanId) ? traceId.lo : parentId;
+    this.spanId = spanId;
+    this.flags = flags;
+  }
+
+  /** Deserializes this from a big-endian byte array */
+  public static TraceContext fromBytes(byte[] bytes) {
+    checkNotNull(bytes, "bytes");
+    if (bytes.length != 32 && bytes.length != 40) {
+      throw new IllegalArgumentException("bytes.length " + bytes.length + " != 32 or 40");
+    }
+
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    long spanId = buffer.getLong(0);
+    long parentId = buffer.getLong(8);
+    TraceId traceId;
+    long flags;
+    if (bytes.length == 32) {
+      long lo = buffer.getLong(16);
+      traceId = new TraceId(0, (parentId == lo) ? parentId : lo);
+      flags = buffer.getLong(24);
+    } else {
```

Comment:
PS here, we are doing exactly the same encoding as before, except longer ID field. We *could* make this a different format, if we like. just seems more straightforward to widen the traceId field for starters..